### DISCUSSION TITLE
Using AirflowSkipException in a PythonVirtualenvOperator

### DISCUSSION THREAD

**Discussion by darrelphamview** [apache/airflow#23921] [Q&A]
Hi guys,

I'm trying to use AirflowSkipException inside a PythonVirtualenvOperator, but I'm having problems with importing it. I'm not too familiar with Python and I understand that the Python Virtualenv creates a totally isolated environment, but I want to skip a task under certain conditions. I'd welcome any suggestions that will help me skip a task using the PythonVirtualenvOperator.

Airflow v2.2.5
Python 3.7
```
from airflow.decorators import dag, task
from airflow.operators.python import PythonVirtualenvOperator

def task1(**kwargs):
    from airflow.exceptions import AirflowSkipException

    somethingWentWrong = True

    if somethingWentWrong is True:
        raise AirflowSkipException("Would like to skip now")



@dag(
    "some dag",
    schedule_interval="* * * * *",
)
def dag1():

    t1 = PythonVirtualenvOperator(
            task_id="Some task", 
            requirements=[],
            system_site_packages=True,
            python_callable=task1,
            op_kwargs="",
            do_xcom_push=False,
            python_version="3.7",
            provide_context=True,
        )

dag = dag1()

__________________________________
[2022-05-25, 19:27:30 UTC] {process_utils.py:168} INFO -   File "/tmp/venv269gtcxw/script.py", line 27, in processDeviceData
[2022-05-25, 19:27:30 UTC] {process_utils.py:168} INFO -     from ***.exceptions import AirflowSkipException
[2022-05-25, 19:27:30 UTC] {process_utils.py:168} INFO - ModuleNotFoundError: No module named '***'
```

I found this discussion https://github.com/apache/airflow/discussions/16661, and it seems like @[rodrigomocca](https://github.com/rodrigomocca) is able to import and use the AirflowFailException.

**Comment by potiuk** [comment]
If you use different python version than the one Airlfow runs on, you are not able to do it out-of-the-box as airflow will be missing in the virtualenv - as you already noticed.

But you can releatively easily implement it by extending PythonVirtualenvOperator and override it's execute() method and try/except CalledProcessError around it (and use exit code from the Python callable) I believe PythonVirtualenv simply return the python callable in a new interpreter, so if you `sys.exit(EXIT_CODE)` in your python callable, it should be available as returncode in the exception and you can decide that some exit_code should be turned into AirflowSkipException.

Just look at the PVO code - it's not that complex. Airflow is all about Python and extending operators rather than treating them as black-boxes is perfectly fine.

**Reply by darrelphamview** [reply]
Thanks. I appreciate the guidance.

**Reply by ayqazi** [reply]
Hi, sorry to revive such an old thread, but did you come up with a solution to this?
